### PR TITLE
hotfix migration script, to remove clusterNetworks and correct

### PR DESCRIPTION
**Pull request by tjjh89017**
bonding slave config

related: https://github.com/harvester/harvester/issues/3045
https://github.com/harvester/harvester/issues/3047

Signed-off-by: Date Huang <date.huang@suse.com>

### CODE REVIEW

**Comment by tjjh89017** [comment]
about this line `$(echo "$IFCFG" | grep "BONDING_SLAVE_" | sed 's/^/ /')`
we need to care for yaml indent.
and that line might be multiple lines.
so every line needs 4 space indent

File: package/harvester-os/files/usr/sbin/migration/migrate_harv_mgmt_to_mgmt_br.sh [hunk 1]
```diff
@@ -39,6 +39,9 @@ function migrate_mgmt_config () {
 	fi
 
 	# start patch
+	# remove cluster network
+	yq -i 'del( .stages.initramfs[0].files[] | select(.path == "*21-harvester-clusternetworks.yaml*"))' "$HARV_CONFIG"
```
[the remark this answers is not in the document]
Comment (tjjh89017) [reply package/harvester-os/files/usr/sbin/migration/migrate_harv_mgmt_to_mgmt_br.sh:43]: Yes, we should.
Do I need to remove those code in this PR?

[the remark this answers is not in the document]
Comment (guangbochen) [reply package/harvester-os/files/usr/sbin/migration/migrate_harv_mgmt_to_mgmt_br.sh:43]: a separate commit looks good enough to me, thanks.

[the remark this answers is not in the document]
Comment (tjjh89017) [reply package/harvester-os/files/usr/sbin/migration/migrate_harv_mgmt_to_mgmt_br.sh:43]: Ok got it.

[the remark this answers is not in the document]
Comment (tjjh89017) [reply package/harvester-os/files/usr/sbin/migration/migrate_harv_mgmt_to_mgmt_br.sh:43]: Fixed

File: package/harvester-os/files/usr/sbin/migration/migrate_harv_mgmt_to_mgmt_br.sh [hunk 2]
```diff
@@ -129,7 +132,7 @@ function migrate_mgmt_config () {
     BONDING_MASTER='yes'
     BOOTPROTO='none'
     POST_UP_SCRIPT="wicked:setup_bond.sh"
-    $(echo "$IFCFG" | grep "BONDING_SLAVE_")
+$(echo "$IFCFG" | grep "BONDING_SLAVE_" | sed 's/^/    /')
```
[the remark this answers is not in the document]
Comment (tjjh89017) [reply package/harvester-os/files/usr/sbin/migration/migrate_harv_mgmt_to_mgmt_br.sh]: I replaced all echo which worked with pipe and grep to "printf".
Fixed

File: package/harvester-os/files/usr/sbin/migration/migrate_harv_mgmt_to_mgmt_br.sh [hunk 3]
```diff
@@ -129,10 +132,10 @@ function migrate_mgmt_config () {
     BONDING_MASTER='yes'
     BOOTPROTO='none'
     POST_UP_SCRIPT="wicked:setup_bond.sh"
-    $(echo "$IFCFG" | grep "BONDING_SLAVE_")
-    $(echo "$IFCFG" | grep "BONDING_MODULE_OPTS")
+$(printf "%s\n" "$IFCFG" | grep "BONDING_SLAVE_" | sed 's/^/    /')
```
[the remark this answers is not in the document]
Comment (tjjh89017) [reply package/harvester-os/files/usr/sbin/migration/migrate_harv_mgmt_to_mgmt_br.sh:135]: We cannot apply this
We should ensure every line will have 4 space indent

If we change it as you mentioned, only first line has 4 space indent' others will not.

This is the root cause of this issue

File: package/harvester-os/files/usr/sbin/migration/migrate_harv_mgmt_to_mgmt_br.sh [hunk 4]
```diff
@@ -29,16 +29,19 @@ function migrate_mgmt_config () {
 	done
 
 	# check DHCP or static
-	if echo "$IFCFG" | grep -q "BOOTPROTO='static'"; then
+	if printf "%s\n" "$IFCFG" | grep -q "BOOTPROTO='static'"; then
```
[the remark this answers is not in the document]
Comment (tjjh89017) [reply package/harvester-os/files/usr/sbin/migration/migrate_harv_mgmt_to_mgmt_br.sh:32]: I will suggest to use printf or echo here for readability

**Comment by bk201** [comment]
```
if ! grep -q "$HARV_MGMT" "$HARV_CONFIG"; then
	echo "$HARV_MGMT not found. Skipping."
	exit 0
fi
```
Is there a better way to check if the config file is already processed? Imagine there is a harvester-mgmt string in any write_files fields.

I'd also suggest adding some tests to the migration script. To collect some common-use-case configuration files, run the script, and check outputs.

**Comment by tjjh89017** [comment]
> Is there a better way to check if the config file is already processed? Imagine there is a harvester-mgmt string in any write_files fields.

If we used "/etc/sysconfig/network/ifcfg-harvester-mgmt" as pattern, it might be better than current one.

**Comment by bk201** [comment]
@tjjh89017 Is it possible to replace the search loop with yq/select?
```
yq '.stages.initramfs[0].files[] | select(.path == "/etc/sysconfig/network/ifcfg-harvester-mgmt") | .content ' /oem/99_custom.yaml
```

**Comment by tjjh89017** [comment]
> @tjjh89017 Is it possible to replace the search loop with yq/select?
> 
> ```
> yq '.stages.initramfs[0].files[] | select(.path == "/etc/sysconfig/network/ifcfg-harvester-mgmt") | .content ' /oem/99_custom.yaml
> ```

Yes, we could.
I changed the function both.
Pleae check it again.
Thanks a lot

File: package/harvester-os/files/usr/sbin/migration/migrate_harv_mgmt_to_mgmt_br.sh [hunk 5]
```diff
@@ -5,40 +5,31 @@ HARV_MGMT="harvester-mgmt"
...
-
-		if yq ".stages.initramfs[0].files[$i].path" "$HARV_CONFIG" | grep -q "/etc/sysconfig/network/ifroute-${HARV_MGMT}"; then
-			# save config
-			IFROUTE=$(yq ".stages.initramfs[0].files[$i].content" "$HARV_CONFIG")
-			echo "$IFROUTE"
-		fi
-	done
+	IFCFG=$(yq '.stages.initramfs[0].files[] | select(.path == "/etc/sysconfig/network/ifcfg-'${HARV_MGMT}'") | .content' "$HARV_CONFIG")
+	IFROUTE=$(yq '.stages.initramfs[0].files[] | select(.path == "/etc/sysconfig/network/ifroute-'${HARV_MGMT}'") | .content' "$HARV_CONFIG")
+	echo "$IFCFG"
```
Comment (Vicente-Cheng) [inline package/harvester-os/files/usr/sbin/migration/migrate_harv_mgmt_to_mgmt_br.sh:16]: I thought that should use `printf "%s\n" "$IFCFG"`? That would be consistent with later operations.

File: package/harvester-os/files/usr/sbin/migration/migrate_harv_mgmt_to_mgmt_br.sh [hunk 6]
```diff
@@ -5,40 +5,31 @@ HARV_MGMT="harvester-mgmt"
...
+	echo "$IFROUTE"
```
Comment (Vicente-Cheng) [inline package/harvester-os/files/usr/sbin/migration/migrate_harv_mgmt_to_mgmt_br.sh:17]: ditto

**Review by Vicente-Cheng — COMMENTED** [review]
Looks good. Just personal opinion, if there have some unit tests (like config before convert and what it should be after convert) may be helpful.

File: package/harvester-os/files/usr/sbin/migration/migrate_harv_mgmt_to_mgmt_br.sh (hunk 5 above)
Comment (tjjh89017) [reply package/harvester-os/files/usr/sbin/migration/migrate_harv_mgmt_to_mgmt_br.sh]: fix

File: package/harvester-os/files/usr/sbin/migration/migrate_harv_mgmt_to_mgmt_br.sh (hunk 6 above)
Comment (tjjh89017) [reply package/harvester-os/files/usr/sbin/migration/migrate_harv_mgmt_to_mgmt_br.sh]: fix

**Review by Vicente-Cheng — APPROVED** [review]
LGTM, please squash commits before merge. Thanks!

**Review by guangbochen — APPROVED** [review]
LGTM, thanks.

**Review by bk201 — APPROVED** [review]
LGTM, thanks.

**Pull request merged**